# Fake News Classifier

###### Importing required libraries

In [5]:
import numpy as np
import pandas as pd

In [6]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


###### Reading data from csv files 

In [7]:
df1 = pd.read_csv('/content/drive/MyDrive/News classification.zip (Unzipped Files)/News classification/True.csv/True.csv')
df2 = pd.read_csv('/content/drive/MyDrive/News classification.zip (Unzipped Files)/News classification/Fake.csv/Fake.csv')
df1['label']='TRUE'
df2['label']='FAKE'

In [8]:
print(df1.shape)
df1.head()

(21417, 5)


title  \
0  As U.S. budget fight looms, Republicans flip t...   
1  U.S. military to accept transgender recruits o...   
2  Senior U.S. Republican senator: 'Let Mr. Muell...   
3  FBI Russia probe helped by Australian diplomat...   
4  Trump wants Postal Service to charge 'much mor...   

                                                text       subject  \
0  WASHINGTON (Reuters) - The head of a conservat...  politicsNews   
1  WASHINGTON (Reuters) - Transgender people will...  politicsNews   
2  WASHINGTON (Reuters) - The special counsel inv...  politicsNews   
3  WASHINGTON (Reuters) - Trump campaign adviser ...  politicsNews   
4  SEATTLE/WASHINGTON (Reuters) - President Donal...  politicsNews   

                 date label  
0  December 31, 2017   TRUE  
1  December 29, 2017   TRUE  
2  December 31, 2017   TRUE  
3  December 30, 2017   TRUE  
4  December 29, 2017   TRUE

In [9]:
print(df2.shape)
df2.head()

(23481, 5)


title  \
0   Donald Trump Sends Out Embarrassing New Year’...   
1   Drunk Bragging Trump Staffer Started Russian ...   
2   Sheriff David Clarke Becomes An Internet Joke...   
3   Trump Is So Obsessed He Even Has Obama’s Name...   
4   Pope Francis Just Called Out Donald Trump Dur...   

                                                text subject  \
0  Donald Trump just couldn t wish all Americans ...    News   
1  House Intelligence Committee Chairman Devin Nu...    News   
2  On Friday, it was revealed that former Milwauk...    News   
3  On Christmas day, Donald Trump announced that ...    News   
4  Pope Francis used his annual Christmas Day mes...    News   

                date label  
0  December 31, 2017  FAKE  
1  December 31, 2017  FAKE  
2  December 30, 2017  FAKE  
3  December 29, 2017  FAKE  
4  December 25, 2017  FAKE

###### Combining the two datasets containing Real and Fake News into one dataset

In [10]:
df = pd.concat([df1, df2])

In [11]:
print(df.shape)
df.describe()

(44898, 5)


title   text  \
count                                               44898  44898   
unique                                              38729  38646   
top     Factbox: Trump fills top jobs for his administ...          
freq                                                   14    627   

             subject                date  label  
count          44898               44898  44898  
unique             8                2397      2  
top     politicsNews  December 20, 2017    FAKE  
freq           11272                 182  23481

In [12]:
df.info

<bound method DataFrame.info of                                                    title  \
0      As U.S. budget fight looms, Republicans flip t...   
1      U.S. military to accept transgender recruits o...   
2      Senior U.S. Republican senator: 'Let Mr. Muell...   
3      FBI Russia probe helped by Australian diplomat...   
4      Trump wants Postal Service to charge 'much mor...   
...                                                  ...   
23476  McPain: John McCain Furious That Iran Treated ...   
23477  JUSTICE? Yahoo Settles E-mail Privacy Class-ac...   
23478  Sunnistan: US and Allied ‘Safe Zone’ Plan to T...   
23479  How to Blow $700 Million: Al Jazeera America F...   
23480  10 U.S. Navy Sailors Held by Iranian Military ...   

                                                    text       subject  \
0      WASHINGTON (Reuters) - The head of a conservat...  politicsNews   
1      WASHINGTON (Reuters) - Transgender people will...  politicsNews   
2      WASHINGTON (Reuter

###### Cleaning the data by dropping columns that aren't required

In [13]:
df.drop(['date','subject'],inplace=True,axis =1)

In [14]:
df.describe

<bound method NDFrame.describe of                                                    title  \
0      As U.S. budget fight looms, Republicans flip t...   
1      U.S. military to accept transgender recruits o...   
2      Senior U.S. Republican senator: 'Let Mr. Muell...   
3      FBI Russia probe helped by Australian diplomat...   
4      Trump wants Postal Service to charge 'much mor...   
...                                                  ...   
23476  McPain: John McCain Furious That Iran Treated ...   
23477  JUSTICE? Yahoo Settles E-mail Privacy Class-ac...   
23478  Sunnistan: US and Allied ‘Safe Zone’ Plan to T...   
23479  How to Blow $700 Million: Al Jazeera America F...   
23480  10 U.S. Navy Sailors Held by Iranian Military ...   

                                                    text label  
0      WASHINGTON (Reuters) - The head of a conservat...  TRUE  
1      WASHINGTON (Reuters) - Transgender people will...  TRUE  
2      WASHINGTON (Reuters) - The special counsel 

###### As we concatenated two datasets, we need to shuffle Real and Fake data

In [15]:
df= df.sample(frac=1)

In [16]:
df.describe

<bound method NDFrame.describe of                                                    title  \
11139  HUMA ABEDIN In Talks To Write “Tell All” For A...   
13101   German police raid flats in hunt for G20 rioters   
11803  Zimbabwe's Mnangagwa promises zero tolerance i...   
21069  PROMINENT DEMOCRAT CLAIMS Old, White, Leftist ...   
16227  WHITE HOUSE LEAKER Reportedly A #NeverTrumper,...   
...                                                  ...   
9719   MELANIA RESPONDS To Ivana’s Claim That SHE’S R...   
10864  UPDATE: UNDER PRESSURE? MISS USA Just Flip Flo...   
14636  VIRAL VIDEO: Bernie Sanders Shoots Hillary The...   
10258  REPORT: TRUMP Has Chosen His Next Communicatio...   
22895  James Comey’s Legacy: Blaming Russia Rather Th...   

                                                    text label  
11139  Huma Abedin is ready to tell all   for a cool ...  FAKE  
13101  BERLIN (Reuters) - Police raided apartments ac...  TRUE  
11803  HARARE (Reuters) - Zimbabwean President Emm

In [17]:
labels = df.label
labels.head()

11139    FAKE
13101    TRUE
11803    TRUE
21069    FAKE
16227    FAKE
Name: label, dtype: object

In [18]:
df.head()

title  \
11139  HUMA ABEDIN In Talks To Write “Tell All” For A...   
13101   German police raid flats in hunt for G20 rioters   
11803  Zimbabwe's Mnangagwa promises zero tolerance i...   
21069  PROMINENT DEMOCRAT CLAIMS Old, White, Leftist ...   
16227  WHITE HOUSE LEAKER Reportedly A #NeverTrumper,...   

                                                    text label  
11139  Huma Abedin is ready to tell all   for a cool ...  FAKE  
13101  BERLIN (Reuters) - Police raided apartments ac...  TRUE  
11803  HARARE (Reuters) - Zimbabwean President Emmers...  TRUE  
21069  The  party of diversity  waits for Hillary to ...  FAKE  
16227  Sunday, GotNews reported that the primary sour...  FAKE

###### Importing required libraries

In [19]:
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier

###### Splitting the data into training and test set

In [20]:
X_train, X_test, y_train, y_test = train_test_split(df.text,labels, test_size=0.2, random_state =5)

###### Initialise Tdidf vectorizer with stopwords and fit the vectorizer with training data and transform it using test data

In [21]:
tfidfv = TfidfVectorizer(stop_words = 'english', max_df =0.5)
tfidf_train = tfidfv.fit_transform(X_train)
tfidf_test = tfidfv.transform(X_test)

In [22]:
classifier = PassiveAggressiveClassifier(max_iter = 50)

In [23]:
classifier.fit(tfidf_train, y_train)

PassiveAggressiveClassifier(max_iter=50)

In [24]:
y_pred = classifier.predict(tfidf_test)

In [25]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score

###### Measuring Accuracy score of the model

In [26]:
accuracy = accuracy_score(y_test,y_pred)
percentAccuracy= round(accuracy*100,2)
print(f'{percentAccuracy}%')

99.34%


> The percentage accuracy of this model is 99.34%

###### Making a confusion matrix to analyse and compare teh given and predicted values

In [27]:
cm = confusion_matrix(y_test,y_pred)
print(cm)

[[4652   31]
 [  28 4269]]


> This model gives 4652 true positives, 4269 true negatives, 28 false positives and 31 false negatives. 